### PR TITLE
[bounty] transformative Crossbreeds (black)

### LINKED ISSUE #1549: Finish Xenobiology Slime Crossbreeds: Black Dominant Crossbreeds
Join this discord server if you're interested in helping with this bounty:  http://discord.gg/Q32uPCS

This exists as an extension of issue #1290 to apply a bounty to a specific family

### PR DESCRIPTION
**Pull request by Evankhell561**
closes #1549
## About The Pull Request
see changelog

## Why It's Good For The Game
More crossbreeds = good

## Changelog
:cl:
add: Dominant Black Extracts!
add: Grey - Slimes split into one additional slime
add: Orange - Slime ignites players with one stack of fire when electrocuting them
add: Purple - Slime regenerates HP faster
add: Blue - One original color slime will always remain after splitting
add: Metal - Slime has 30% extra max HP
add: Yellow - Slime gains electric charge much faster
add: Dark Purple - Slime rapidly converts atmospheric plasma to oxygen, healing in the process.
add: Dark Blue - Slime takes greatly reduced damage from exposure to water (50% less)
add: Silver - Slime no longer loses nutrition over time, and is no longer able to feed.
add: Bluespace - Slimes will teleport to targets when they are at full electric charge.
add: Sepia - Slime moves 30% faster
add: Cerulean - Slime splits into two identical adult slimes with half hunger when it reproduces. Always 0% mutation chance.
add: Pyrite - Slime always splits into totally random colors, equal chance for every slime type except rainbow. Can never yield a rainbow slime.
add: Red - Slime does 10% more damage when feeding and attacking
add: Green - Slime can completely consume corpses after death for a little extra nutrition
add: Pink - Slime gains and will prefer to speak in common.
add: slime extracts can now be sold in cargo.
add: Gold - Slimes may be sold to cargo for extra money.
add: Oil - Slime douses anything it feeds on in welding fuel.
add: Black - Slime is nearly transparent. Slime is fully visible while feeding.
add: Light Pink - Ghosts may possess the slime at will, added to spawners menu.
add:Adamantine - Slime takes greatly reduced damage from brute force attacks (50%)
add: Rainbow - Slime randomly changes color periodically. Reproduction follows mutation chances of the current color
/:cl:

### CODE REVIEW

**Comment by ZombiesVsPlants** [comment]
jesus good job

**Comment by leestriter** [comment]
Amazing. I've been hoping to see this for a long time

**Comment by MNarath1** [comment]
are the effects stackable?

**Comment by Evankhell561** [comment]
> are the effects stackable?

no.

**Comment by knacker48** [comment]
> > are the effects stackable?
> 
> no.

Damn, I was gonna make a ultra death slime and make it my bodyguard

File: code/game/objects/items/devices/scanners.dm
Comment (PowerfulBacon) [inline code/game/objects/items/devices/scanners.dm]: There has got to be a better way to do it than this, seems kinda shitcodey

Comment (PowerfulBacon) [reply code/game/objects/items/devices/scanners.dm]: Also does that I thing even work? You could make a list and do list.join instead

File: code/modules/mob/living/simple_animal/slime/life.dm
Comment (PowerfulBacon) [inline code/modules/mob/living/simple_animal/slime/life.dm:13]: Don't do it on life if it's a 1 time thing that needs changing

File: code/game/objects/items/devices/scanners.dm
Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm]: I know I just kept this here in case in the future someone wanted to make affections stackable. I will change it.

Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm]: ready.

**Comment by MNarath1** [comment]
> > are the effects stackable?
> 
> 
> 
> no.

Just to be sure do you mean not stackable with the same effects or not stackable with different effects?

**Comment by Rukofamicom** [comment]
Just finished extensively testing this.

**Working as Intended:**
I tested every transformative extract, and every color is passed on to offspring as intended. Additionally using a new transformative extract overwrites the old one's effect. Transformative Grey, Orange, Metal, Dark Blue, Silver, Sepia, Cerulean, Pyrite, Red, Green, Gold, Oil, Adamantine and Rainbow extracts are all working exactly as described with no bugs that I can find. _I couldn't really discern whether Purple was working or not_

**Working, but suggested tweaks** (Optional to fix)
- Blue - prevents rainbow slimes from forming, even with 100% mutation chance. Considering the intended effect of blue, this can be considered a side-effect rather than an outright bug and I'm not hard-pressed to see it fixed. 
- Yellow - Could use a buff in the charge generation, since it's not that much of an increase currently. I would suggest yellow double charge generation since this is an extract that can only be used on sentient slimes; NPC slimes will never shock or tackle targets. 
- Dark Purple - Definitely working, but again it's not terribly strong. Healing is only noticeable in high concentrations of plasma, and since the slime drains a percentage of the available plasma it is also never going to completely purify the air. I'd suggest making it drain a set amount of plasma from the nearby atmos every time it activates rather than a percentage of available plasma.
- Bluespace - Add a sound effect and/or sparks to the teleport. It is currently completely silent and the slime simply appears on top of their target. 
- Black - Slimes are still very visible as opposed to barely visible. Suggest roughly doubling the transparency. 

**Bugs** (Should be fixed before merge)
- Dark purple - Spews debug text constantly referring to the amount of plasma that has been consumed (This is how I knew it was percentage based :laughing:)
- Bluespace - Cannot be used by sentient slimes - if this can't be easily fixed it's probably fine being NPC-only, given how powerful it would be in player hands. If maintainers agree that it's fine, then I'm fine with it for sake of bounty.
- Pink - Does not work for NPC slimes, they still speak in slime even though they have common. 
- Light Pink - Spawner menu only shows the slime that the extract was applied to, and shows it by exact name Grey slime (234). Once that slime reproduces, the offspring are not added to spawners menu and the existing spawner is bugged. If possible, I'd suggest adding them to the spawn menu as generic "Xenobiology Slime".

**Comment by Rukofamicom** [comment]
All of my concerns and tweaks in the above post have been resolved, although I have one more _potential_ bug that will require testmerge (or at least two people on private) to check. 

Transformative/Light pink mutated slimes **might** remain possible to possess by ghosts even after they already house a soul. I say this because when I aghost out of one, the spawner still presents the slime I ghosted out of as a valid option. 

That potential problem aside, this is ready for a merge now

**Comment by Rukofamicom** [comment]
Someone showed me a trick to testing this by myself: 

Transformative/Light mutated pink slimes continue to show up in the spawners menu for ghosts even after a ghost has possessed them, however it does **not** allow for players to assume control of the mob and kick the existing player out.

**Comment by leestriter** [comment]
Excited to see this merged!

File: code/modules/mob/living/simple_animal/slime/powers.dm
Comment (PowerfulBacon) [inline code/modules/mob/living/simple_animal/slime/powers.dm]: Is this not setting the language of the slime making the new slime, not the baby slime?

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (PowerfulBacon) [inline code/modules/mob/living/simple_animal/slime/slime.dm]: Since this is added to the list by:
`LAZYADD(GLOB.mob_spawners["[master.real_name]'s slime"], M)`
Can't you remove by just removing from
`LAZYREMOVE(GLOB.mob_spawners["[master.real_name]'s slime"], M)`

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (PowerfulBacon) [inline code/modules/mob/living/simple_animal/slime/slime.dm]: `humanize_slime` is kind of a bad name for it, it implies you are turning the slime into a human not having a ghost assume control of the slime.

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (PowerfulBacon) [inline code/modules/research/xenobiology/crossbreeding/transformative.dm]: Applying another slime core on something that already has this one will remove the old effect?
Applying bluespace, then yellow removes the teleport verb?

File: code/modules/research/xenobiology/xenobiology.dm
Comment (PowerfulBacon) [inline code/modules/research/xenobiology/xenobiology.dm]:
```suggestion
	if (sparkly)
```

File: code/game/objects/items/devices/scanners.dm
Comment (PowerfulBacon) [inline code/game/objects/items/devices/scanners.dm]: If there is a way of doing it without the if, if, if, if then that would be great.
If not remove i and have a list instead and add the slime effect name to the list each time then in the to_chat do list.join(",")
["blue", "pink"].join(", ") = "blue, pink"

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (Evankhell561) [reply code/modules/research/xenobiology/crossbreeding/transformative.dm]: > Applying another slime core on something that already has this one will remove the old effect?
Applying bluespace, then yellow removes the teleport verb?

yes

File: code/modules/mob/living/simple_animal/slime/life.dm
Comment (Evankhell561) [reply code/modules/mob/living/simple_animal/slime/life.dm:13]: this value is changed multiple times

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (Rukofamicom) [reply code/modules/research/xenobiology/crossbreeding/transformative.dm]: Yes it does, this is one of the things I tested for - though I didn't try every single combination of slime, I did check that each of them loses the effect when a new extract is applied.

**Comment by leestriter** [comment]
Still excited for this.

File: code/game/objects/items/devices/scanners.dm
Comment (PowerfulBacon) [inline code/game/objects/items/devices/scanners.dm:753]: now this just shows as slime effects detector: greyredbluerainbow
you need something like
var/list/effects = list()
effects += "black"
effects += "rainbow"]
if(effects.len)
   to_chat(user, "Slime effects detected: [effects.join(", ")]")

Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm:753]: a slime can only have one effect at a time, I don't see the problem.

Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm:753]: > now this just shows as slime effects detector: greyredbluerainbow

Where did you find out what is wrong?

Comment (Rukofamicom) [reply code/game/objects/items/devices/scanners.dm:753]: I mirror this - where are you getting this error? When I examine slimes with the console it always only shows me one color no matter how many times I've changed the color.

Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm:753]: > a slime can only have one effect at a time

Comment (PowerfulBacon) [reply code/game/objects/items/devices/scanners.dm:753]: I didn't know that

Comment (PowerfulBacon) [reply code/game/objects/items/devices/scanners.dm:753]: But why is it using bit flags if you can only have 1 effect at once?

Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm:753]: initially I wanted to facilitate the work of a possible rework where someone wants to make stackable effects.

Comment (leestriter) [reply code/game/objects/items/devices/scanners.dm:753]: So, are you gonna redo it the way Bacon said? I'd love to see this merged it's super cool.

**Comment by leestriter** [comment]
Hope this gets merged!

File: code/game/objects/items/devices/scanners.dm
Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm:753]: > So, are you gonna redo it the way Bacon said?

no

Comment (leestriter) [reply code/game/objects/items/devices/scanners.dm:753]: That's too bad. Thanks for making it, sadly don't think it'll get merged in that case.

Comment (Evankhell561) [reply code/game/objects/items/devices/scanners.dm:753]: > That's too bad. Thanks for making it, sadly don't think it'll get merged in that case.

I don't see why it shouldn't be marged. Your suggestion was made by having the wrong thought about what should happen.

He made that suggestion to correct an error but the error never existed.

Comment (leestriter) [reply code/game/objects/items/devices/scanners.dm:753]: Ah ok my bad.

File: code/__DEFINES/mobs.dm
Comment (park66665) [inline code/__DEFINES/mobs.dm]:
```suggestion
#define SLIME_EFFECT_DEFAULT		(1<<0)
#define SLIME_EFFECT_GREY			(1<<1)
#define SLIME_EFFECT_ORANGE			(1<<2)
#define SLIME_EFFECT_PURPLE			(1<<3)
#define SLIME_EFFECT_BLUE			(1<<4)
#define SLIME_EFFECT_METAL			(1<<5)
#define SLIME_EFFECT_YELLOW			(1<<6)
#define SLIME_EFFECT_DARK_PURPLE	(1<<7)
#define SLIME_EFFECT_DARK_BLUE		(1<<8)
#define SLIME_EFFECT_SILVER			(1<<9)
#define SLIME_EFFECT_BLUESPACE		(1<<10)
#define SLIME_EFFECT_SEPIA			(1<<11)
#define SLIME_EFFECT_CERULEAN		(1<<12)
#define SLIME_EFFECT_PYRITE			(1<<13)
#define SLIME_EFFECT_RED			(1<<14)
#define SLIME_EFFECT_GREEN			(1<<15)
#define SLIME_EFFECT_PINK			(1<<16)
#define SLIME_EFFECT_GOLD			(1<<17)
#define SLIME_EFFECT_OIL			(1<<18)
#define SLIME_EFFECT_BLACK			(1<<19)
#define SLIME_EFFECT_LIGHT_PINK		(1<<20)
#define SLIME_EFFECT_ADAMANTINE		(1<<21)
#define SLIME_EFFECT_RAINBOW		(1<<22)
```

I don't think `//done` comments are necessary now.

File: code/modules/cargo/exports/xenobio.dm
Comment (park66665) [inline code/modules/cargo/exports/xenobio.dm]: I think those colors are a bit underpriced. gold, green, red, and pink slimes are quite higher up in the tiers; they are generally as common as colors beneath: silver, dark blue, dark purple, and yellow.

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/slime.dm:550]: The text you put as a `flavor_text` sounds more like a `short_desc`: You are that thing etc etc. Take a look at `/obj/effect/mob_spawn`'s `short_desc` and `flavor_text`.

File: code/modules/mob/living/simple_animal/slime/powers.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/powers.dm]:
```suggestion
			maxHealth = 200
			if(transformeffects & SLIME_EFFECT_METAL)
				maxHealth = round(maxHealth * 1.3)
```

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/slime.dm]:
```suggestion
```

This var currently is not in use.

File: code/modules/mob/living/simple_animal/slime/powers.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/powers.dm]:
```suggestion
```

ditto

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (park66665) [inline code/modules/research/xenobiology/crossbreeding/transformative.dm]:
```suggestion
```

ditto

File: code/modules/mob/living/simple_animal/slime/powers.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/powers.dm]:
```suggestion
		M.add_verb(/mob/living/simple_animal/slime/proc/teleport)
```

See #2771.

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (park66665) [inline code/modules/research/xenobiology/crossbreeding/transformative.dm]:
```suggestion
		S.remove_verb(/mob/living/simple_animal/slime/proc/teleport)
```

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (park66665) [inline code/modules/research/xenobiology/crossbreeding/transformative.dm]:
```suggestion
	S.add_verb(/mob/living/simple_animal/slime/proc/teleport)
```

File: code/modules/research/xenobiology/xenobiology.dm
Comment (park66665) [inline code/modules/research/xenobiology/xenobiology.dm]:
```suggestion
	if(sparkly)
```

File: code/modules/surgery/core_removal.dm
Comment (park66665) [inline code/modules/surgery/core_removal.dm]:
```suggestion
		if(slime.transformeffects & SLIME_EFFECT_GOLD)
```

File: code/modules/mob/living/simple_animal/slime/life.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/life.dm]: If the slime in question doesn't barf its contents on its death (I haven't tested it but slimes probably won't; mobs having other mobs as contents generally requires special handling; see morphs and laughter demons), then this is a bit of a very weird situation. You also can call `dust`, `gib`, or `qdel` if you want to cause permadeath.

**Review by park66665 — COMMENTED** [review]
I read through code changes rather quickly. Feel free to ask. (24/7 availability NOT guaranteed)

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (Evankhell561) [reply code/modules/research/xenobiology/crossbreeding/transformative.dm]: image

File: code/modules/mob/living/simple_animal/slime/powers.dm
Comment (park66665) [reply code/modules/mob/living/simple_animal/slime/powers.dm]: Try to re-apply this series of changes.

File: code/modules/research/xenobiology/crossbreeding/transformative.dm
Comment (park66665) [reply code/modules/research/xenobiology/crossbreeding/transformative.dm]: Update your branch.

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (utilbird) [inline code/modules/mob/living/simple_animal/slime/slime.dm:539]: There's already a key check above, I don't think this is needed

**Review by utilbird — APPROVED** [review]

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (Evankhell561) [reply code/modules/mob/living/simple_animal/slime/slime.dm:539]: There is the possibility that a second player enters the slime while the first still decides to answer the question of becoming a slime.

Comment (utilbird) [reply code/modules/mob/living/simple_animal/slime/slime.dm:539]: Oh, right. Xd

File: code/modules/mob/living/simple_animal/slime/powers.dm
Comment (utilbird) [inline code/modules/mob/living/simple_animal/slime/powers.dm]:
```suggestion
			for(var/i=1, i<=childamount, i++)
```

**Review by utilbird — DISMISSED** [review]

**Comment by utilbird** [comment]
Oops. didn't mean to approve twice. Enjoy your double tick

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (park66665) [reply code/modules/mob/living/simple_animal/slime/slime.dm:550]: Screenshot 2021-01-27 173807

This thing surely isn't critical, but you now can see the discrepancies.

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/slime.dm:555]:
```suggestion
/mob/living/simple_animal/slime/death(gibbed)
	remove_form_spawner_menu()
	. = ..()

/mob/living/simple_animal/slime/ghostize(can_reenter_corpse = 1)
	. = ..()
	if(. && transformeffects & SLIME_EFFECT_LIGHT_PINK)
		LAZYADD(GLOB.mob_spawners["[S.master.real_name]'s slime"], S)

```

I think spawner menu re-addition should happen when the mob ghosts.

**Comment by ike709** [comment]
This will need wiki changes before it's merged.

File: code/modules/mob/living/simple_animal/slime/slime.dm
Comment (park66665) [inline code/modules/mob/living/simple_animal/slime/slime.dm]:
```suggestion
	GLOB.poi_list |= src
```

**Review by park66665 — APPROVED** [review]
Someone has to YOLOmerge it. I will. YOLOOOOOOOOO

**Pull request merged**
